src/Kestrel.Core/Internal/Infrastructure/TimeoutControlExtensions.cs

```diff
@@ -18,7 +18,8 @@ public static void StartDrainTimeout(this ITimeoutControl timeoutControl, MinDat
             // There's no reason to stop timing the write after the connection is closed.
             var oneBufferSize = maxResponseBufferSize.Value;
             var maxBufferedBytes = oneBufferSize < long.MaxValue / 2 ? oneBufferSize * 2 : long.MaxValue;
-            timeoutControl.StartTimingWrite(minDataRate, maxBufferedBytes);
+            timeoutControl.BytesWritten(minDataRate, maxBufferedBytes);
```

Comment:
Since we now keep track of all bytes written, and extend the write timeout as needed, it's tempting to no longer add 2 times the max buffer size to the bytes written accounting for the connection drain timeout. As we've discussed before, this add several minutes to the timeout with the default 240 bytes/sec rate limit.

I'm thinking instead this line changes to `timeoutControl.BytesWritten(minDataRate, 1);` to add a grace period to the drain if necessary. What do you think @Tratcher?

Reply:
👍